### ISSUE TITLE
Accessibility markup problems in Carbon react overflow menu

### ISSUE THREAD

**Issue by carmacleod** [carbon-design-system/carbon#2495]
In the [Carbon react overflow menu](http://react.carbondesignsystem.com/?selectedKind=OverflowMenu&selectedStory=basic&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybook%2Fstories%2Fstories-panel):
1. the `div` containing the overflow icon needs to have `role="button"`, not `role="menu"`
2. the button needs a nice default label in case one is not provided, say `aria-label="Menu"`
3. the `ul` with `role="menu"` needs to have the same `aria-label` as the button
4. there's a `div` with `role="menuitem"` in the DOM for no apparent reason
(see screen snap, below. This may be the cause of https://github.com/IBM/carbon-components-react/issues/1237)
5. the `div` with `role="button"` (see 1) and the `ul` with `role="menu"` need to be siblings, i.e. the `ul` should immediately follow the button in the DOM, so that screen reader users can use their virtual cursor to read through the menu items without getting totally lost.

![image](https://user-images.githubusercontent.com/3331913/55775888-cb92b980-5a68-11e9-8112-37b7fcb87983.png)

**Comment by emyarod** [comment]
@carmacleod this issue also applies to the vanilla markup right? since the react markup mostly follows the vanilla markup

**Comment by dakahn** [comment]
by @emyarod in https://github.com/carbon-design-system/carbon/pull/2906#pullrequestreview-245450384
> 
> 
> ![image](https://user-images.githubusercontent.com/8265238/58884884-d9e70580-86a6-11e9-8a28-3ed9925cbb0f.png)
> 
> there's a visual bug where the overflow menu overlaps with the trigger button. this might be because the menu body is being rendered in two separate locations. this could also be affecting keyboard navigation
> 
> the aria attribute changes look good to me, but the fix for the screen reader getting lost (making the `ul[role="menu"]` a sibling of the `div[role="button"]`) is not so straightforward as replacing `wrappedMenuBody` with `menuBody` and moving it. instead I think we should modify the target of the `<FloatingMenu>` portal
>

**Comment by asudoh** [comment]
Wrt 5. - Overflow menu tends to be used in table cell and thus putting menu body next to the menu trigger button (in a table cell) causes the menu body's visibility constrained by the table cell's box. Looks that [`aria-owns`](https://www.w3.org/TR/wai-aria-1.1/#aria-owns) may help this problem space.

**Comment by emyarod** [comment]
when taken out of the context of #2906 that quote doesn't make sense since it's referencing changes that were made in the PR

but as @asudoh mentioned, making the menu a sibling to the trigger button can be an issue when the menu is placed inside a container like a table cell. support for non floating menus was recently deprecated in https://github.com/carbon-design-system/carbon-components-react/issues/1921

**Comment by dakahn** [comment]
So where does that leave us? I don't understand what "...I think we should modify the target of the `<FloatingMenu>` portal" means. 😕

**Comment by emyarod** [comment]
we can ignore the quote since it's no longer in context or relevant. right now we can't apply the suggested fix of placing the menu next to the trigger button, so we need to come up with an alternative solution. `aria-owns` as @asudoh mentioned may be a possibility

**Comment by elizabethsjudd** [comment]
WH is actually working on Overflow Menu now for handlebars implementation of Carbon and we've ran in to many of the same issues as logged by @carmacleod but have it working when there is `overflow: hidden` parent.  I have a meeting with our designer today that will hopefully lead to some collaboration on this component based off of my findings.

**Comment by elizabethsjudd** [comment]
wrt item 5: I was speaking with Sharon on IBMa, and when the trigger is a sibling of the menu, the trigger element should also use the `aria-controls` attribute to create the relationship between the two elements.

**Comment by emyarod** [comment]
the main issue we are facing is making the menu a sibling of the trigger button. is there an alternative to that markup structure?

**Comment by elizabethsjudd** [comment]
@emyarod For accessibility I have always had to have the trigger as a sibling to the menu.  I've built multiple working examples of this component that are accessible and work in various different use cases.  The UX is different so maybe if we could set up a meeting between are two teams to discuss?

**Comment by emyarod** [comment]
we recently deprecated our non floating menu which would be used in this scenario, so we have to first reexamine our decision to deprecate it if there is no suitable workaround using the current floating menu

unfortunately I can't link to the relevant issue or PR since it appears it was lost during the monorepo move and/or when the issue tracker was disabled on the old React repo...

**Comment by elizabethsjudd** [comment]
I think getting a meeting together to discuss some of our options would be a good idea.  Is there someone from @carbon-design-system/design to include as well?  I'll look to set something up for later next week.

**Comment by asudoh** [comment]
BTW from https://www.w3.org/TR/wai-aria-1.1/#aria-owns:

> Identifies an element (or elements) in order to define a visual, functional, or contextual parent/child relationship between DOM elements **where the DOM hierarchy cannot be used to represent the relationship**.

**Comment by carmacleod** [comment]
Apologies for my suggestion to use "Menu" as a default `aria-label` when a label is not provided.
Windows screen readers say "[label] menu button" (NVDA) or "[label] button menu" (JAWS) when focus goes to a button with `aria-haspopup="true"`, and so it turns out that "Menu" is redundant (i.e. NVDA says "Menu menu button" and JAWS says "Menu button menu"). A better default string might be "Overflow" or "More...".

Regarding the label for the dropdown ul, it should really use `aria-labelledby` if possible, which can be set to the id for the button, so that the menu always picks up the same label as the button. That would also work if the button had visible text instead of an icon. See the [APG menu button example](http://w3c.github.io/aria-practices/examples/menu-button/menu-button-actions.html) for an example of using `aria-labelledby` on the dropdown menu element.

Regarding the problem of having the menu show when the button is in a table and sibling markup is used, [this example](https://www.codeply.com/go/QnOrPZvtHt/bootstrap-dropdown-overlay-table-cell) seems to have solved that problem.

**Comment by asudoh** [comment]
Thanks @carmacleod for more suggestion!

> Regarding the problem of having the menu show when the button is in a table and sibling markup is used, [this example](https://www.codeply.com/go/QnOrPZvtHt/bootstrap-dropdown-overlay-table-cell) seems to have solved that problem.

By constraining the height of `<div class="table-responsive">` in the link, it demonstrates one of the problems I demonstrated offline - The menu is crippled by the scrolling container.

**Comment by carmacleod** [comment]
> The menu is crippled by the scrolling container.

Ah, I see. I was only looking at the first menu button. :)
